cmd/ebrelayer/relayer/ethereum.go

```diff
@@ -129,243 +139,225 @@ func (sub EthereumSub) Start(txFactory tx.Factory, completionEvent *sync.WaitGro
+			FromBlock: big.NewInt(int64(fromBlockNumber)),
+			ToBlock:   big.NewInt(int64(endBlock)),
+			Addresses: []common.Address{bridgeBankAddress},
+			Topics:    topics,
+		})
 
 		if err != nil {
-			log.Printf("failed to start a client %s\n", err.Error())
-			continue
+			sub.SugaredLogger.Errorw("failed to filter the logs from ethereum client",
+				errorMessageKey, err.Error())
+			return
 		}
 
-		for _, result := range block.TxsResults {
-			for _, event := range result.Events {
-				log.Printf("Replay get an event %s\n", event.GetType())
-				if event.GetType() == "create_claim" {
-					claim, err := txs.AttributesToEthereumBridgeClaim(event.GetAttributes())
-					if err != nil {
-						continue
-					}
-
-					// Check if sender is me
-					if claim.CosmosSender.Equals(sub.ValidatorAddress) {
-						log.Printf("We got a eth bridge claim message %s\n", claim.EthereumSender.String())
-						claimArray = append(claimArray, claim)
-					}
-				}
+		// loop over ethlogs, and build an array of burn/lock events
+		for _, ethLog := range ethLogs {
+			event, isBurnLock, err := sub.logToEvent(oracletypes.NetworkDescriptor(networkID.Uint64()),
+				bridgeBankAddress,
+				bridgeBankContractABI,
+				ethLog)
+
+			if err != nil {
+				sub.SugaredLogger.Errorw("failed to transform from log to event.",
+					errorMessageKey, err.Error())
+				continue
+			}
+			if !isBurnLock {
+				sub.SugaredLogger.Infow("not burn or lock event, continue events.")
+				continue
 			}
+			events = append(events, event)
 		}
-	}
-
-	return claimArray
-}
 
-// EventProcessed check if the event processed by relayer
-func EventProcessed(bridgeClaims []types.EthereumBridgeClaim, event types.EthereumEvent) bool {
-	for _, claim := range bridgeClaims {
-		if event.From == claim.EthereumSender && event.Nonce.Cmp(claim.Nonce.BigInt()) == 0 {
-			return true
+		if len(events) > 0 {
+			var nextLockBurnNonce uint64
+			if nextLockBurnNonce, err = sub.handleEthereumEvent(txFactory, events, symbolTranslator, lockBurnNonce); err != nil {
+				sub.SugaredLogger.Errorw("failed to handle ethereum event.",
+					errorMessageKey, err.Error())
+				return
+			}
+			// handleEthereumEvent return the next expected lock burn nonce
+			lockBurnNonce = nextLockBurnNonce - 1
+			time.Sleep(transactionInterval)
 		}
+
+		// update fromBlockNumber
+		fromBlockNumber += maxQueryBlocks
 	}
-	return false
 }
 
 // Replay the missed events
-func (sub EthereumSub) Replay(txFactory tx.Factory, fromBlock int64, toBlock int64, cosmosFromBlock int64, cosmosToBlock int64, symbolTranslator *symbol_translator.SymbolTranslator) {
-	log.Printf("ethereum replay for %d block to %d block\n", fromBlock, toBlock)
-
-	bridgeClaims := sub.getAllClaims(cosmosFromBlock, cosmosToBlock)
-	log.Printf("found out %d bridgeClaims\n", len(bridgeClaims))
+func (sub EthereumSub) Replay(txFactory tx.Factory, symbolTranslator *symbol_translator.SymbolTranslator) {
```

Comment:
Do we still need Replay?

Reply:
I think replay is unnecessary, will create other pr to remove it completely.